### PR TITLE
fix(preview): resolve indirect object JSON serialization and add prev…

### PR DESCRIPTION
**Pull request by xiez**
…iew retry trigger

### CODE REVIEW

File: backend/documents/tasks.py [hunk 1]
```diff
@@ -89,6 +89,20 @@ def convert_office_to_pdf_task(version_id):
         logger.error(f"Error converting document version {version_id}: {e}")
 
 
+def resolve_pdf_object(obj):
```
Comment (xiez) [inline backend/documents/tasks.py]: 🟡 **[Medium] `resolve_pdf_object` is a private helper but lacks a leading underscore**

This function is an internal PDF-parsing utility but is exported as a public name. The test file imports it directly (`from documents.tasks import resolve_pdf_object`), which locks it into the public API.

Rename to `_resolve_pdf_object` and update the test import accordingly to signal its private nature.

File: backend/documents/tasks.py [hunk 2]
```diff
@@ -89,6 +89,20 @@ def convert_office_to_pdf_task(version_id):
...
+    visited = set()
+    while hasattr(obj, "get_object"):
+        obj_id = id(obj)
+        if obj_id in visited:
```
Comment (xiez) [inline backend/documents/tasks.py:99]: 🟡 **[Medium] Cycle detection uses `id()` — unreliable if intermediate objects are GC'd**

`id()` in CPython returns the memory address of the object. If intermediate objects in the chain are garbage-collected mid-loop, their addresses can be reused by new objects, causing a false cycle detection and prematurely terminating resolution.

This is acceptable as a best-effort guard given pypdf objects may not be hashable, but deserves a comment explaining the limitation:
```python
# NOTE: We use id() as a best-effort cycle guard. CPython may reuse addresses
# for collected objects, but the while-loop's resolved-is-obj identity check
# provides a secondary safety net.
```

File: backend/documents/tasks.py [hunk 3]
```diff
@@ -98,33 +112,43 @@ def _extract_links_for_page(pdf_page, page_num):
...
-            uri = action.get("/URI") if action and action.get("/S") == "/URI" else None
+            action = resolve_pdf_object(obj.get("/A"))
+            rect = resolve_pdf_object(obj.get("/Rect"))
+            uri = None
+            if action and resolve_pdf_object(action.get("/S")) == "/URI":
+                uri_obj = resolve_pdf_object(action.get("/URI"))
+                if isinstance(uri_obj, bytes):
+                    uri = uri_obj.decode("utf-8", errors="ignore")
+                elif uri_obj is not None:
+                    uri = str(uri_obj)
```
Comment (xiez) [inline backend/documents/tasks.py]: 🟢 **[Low] `str(uri_obj)` fallback is too broad**

If `uri_obj` is a pypdf `NameObject` (e.g. `/URI`) or `ArrayObject`, `str()` produces a repr like `'/URI'` or `"['/something']"` — a malformed URL that would be silently stored or skipped.

A more explicit guard:
```python
elif isinstance(uri_obj, str):
    uri = uri_obj
# else: skip non-string, non-bytes URI objects
```

File: backend/documents/views.py [hunk 4]
```diff
@@ -659,6 +660,27 @@ def get(self, request, document_id, *args, **kwargs):
                 status=status.HTTP_404_NOT_FOUND
             )
 
+        force_rebuild = request.query_params.get('force_rebuild') == 'true'
+        if force_rebuild:
```
Comment (xiez) [inline backend/documents/views.py]: 🔴 **[Critical] `force_rebuild` is a destructive state mutation triggered via GET — violates HTTP semantics**

This branch deletes all page rows, resets render fields, deletes physical files from the fileserver, and queues a Celery task — all from a `GET` request.

GET requests must be idempotent per HTTP semantics (RFC 7231). This is unsafe because:
- Browser prefetch and link crawlers can trigger it silently
- GET cannot be protected by CSRF tokens
- Accidental page reloads will destroy in-progress renders

**Fix:** Move this to a dedicated `POST` action, e.g.:
```
POST /api/v1/documents/{id}/rebuild-preview/
```

File: backend/documents/views.py [hunk 5]
```diff
@@ -659,6 +660,27 @@ def get(self, request, document_id, *args, **kwargs):
...
+                
+            # Clean up fileserver storage outside the transaction
+            for storage_key in page_storage_keys:
+                try:
+                    fileserver_client.delete_file(storage_key)
+                except Exception as e:
+                    logger.error(f"Failed to delete page file {storage_key} during force rebuild: {e}")
+
         preview_mode = preview_mode_for_version(primary_version)
         render_status = enqueue_server_preview_render(primary_version)
```
Comment (xiez) [inline backend/documents/views.py:664]: 🔴 **[Critical] No rate limit or cooldown — rapid calls can loop-destroy mid-generation renders**

There is no guard against calling `?force_rebuild=true` in rapid succession. Each call:
1. Deletes pages that may be mid-generation by a currently-running Celery task
2. Resets `has_pages=False`, defeating the `if version.has_pages: return` guard in `generate_pdf_pages_task`
3. Queues a new Celery task

This means a fast loop of requests will spawn unbounded render tasks and continuously abort in-flight work.

**Fix:** Add an explicit guard before triggering rebuild:
```python
if primary_version.render_status == DocumentVersion.RENDER_PROCESSING:
    return Response({'detail': 'Render already in progress.'}, status=400)
```

File: backend/documents/views.py [hunk 6]
```diff
@@ -659,6 +660,27 @@ def get(self, request, document_id, *args, **kwargs):
...
+            # Collect storage keys before deleting rows
+            page_storage_keys = list(primary_version.pages.values_list('storage_key', flat=True))
+            
+            # DB operations wrapped in transaction block
+            with transaction.atomic():
+                primary_version.render_status = DocumentVersion.RENDER_NOT_GENERATED
+                primary_version.render_error = ''
+                primary_version.has_pages = False
+                primary_version.num_pages = None
+                primary_version.save(update_fields=['render_status', 'render_error', 'has_pages', 'num_pages'])
```
Comment (xiez) [inline backend/documents/views.py]: 🟡 **[Medium] Missing `'updated_at'` in `update_fields`**

Every other `version.save()` call in the codebase includes `'updated_at'` in `update_fields` (e.g. in `generate_pdf_pages_task` and `convert_office_to_pdf_task`). This is inconsistent and could silently break if `BaseModel` is ever changed away from `auto_now=True`.

```python
primary_version.save(update_fields=['render_status', 'render_error', 'has_pages', 'num_pages', 'updated_at'])
```

File: frontend/src/components/documents/DocumentPreviewModal.jsx [hunk 7]
```diff
@@ -29,9 +29,19 @@ export function DocumentPreviewModal({ documentId, versionId = null, isOpen, onO
...
   const [zoomLevel, setZoomLevel] = useState(1);
   const [totalPages, setTotalPages] = useState(1);
+  const [rebuildTriggerCount, setRebuildTriggerCount] = useState(0);
+  
+  const handleRetry = () => {
+    setRebuildTriggerCount((prev) => prev + 1);
+  };
+
+  useEffect(() => {
+    setRebuildTriggerCount(0);
```
Comment (xiez) [inline frontend/src/components/documents/DocumentPreviewModal.jsx]: 🟡 **[Medium] `rebuildTriggerCount` reset race — two `useEffect`s with overlapping dependencies**

This effect resets `rebuildTriggerCount` to `0`. Effect B (the fetch effect at L57) also re-runs when `documentId`/`versionId`/`isOpen` change, reading the **stale** counter before this state update is committed.

The guard works incidentally because `isSameContext` is `false` at that moment (the ref hasn't been updated yet), but this is a fragile timing dependency on React's effect ordering and ref mutation order — not an explicit design guarantee.

Consider merging both effects into one, or using `useReducer` to make the reset-and-fetch transition atomic.

File: frontend/src/components/documents/PreviewStatePanel.jsx [hunk 8]
```diff
@@ -28,6 +30,19 @@ export function PreviewStatePanel({
...
+  useEffect(() => {
+    setShowStuckRetry(false);
+    if (!isPending) {
+      return;
+    }
+    const timer = setTimeout(() => {
+      setShowStuckRetry(true);
+    }, 60000); // Show retry if stuck for > 60s
+    return () => clearTimeout(timer);
+  }, [isPending, documentData?.preview_status]);
```
Comment (xiez) [inline frontend/src/components/documents/PreviewStatePanel.jsx:47]: 🟡 **[Medium] Timer dependency on both `isPending` and `documentData?.preview_status` is correct but undocumented**

`isPending` is derived from `documentData?.preview_status`, so including both in the dependency array means the 60-second timer resets whenever the status changes — even when toggling between two pending states (`not_generated` → `processing`). This is the correct UX behaviour.

However, a future maintainer may simplify the deps to just `[isPending]`, breaking the per-status-change reset. Please add a comment:
```js
// Both isPending and preview_status are listed as deps intentionally:
// the timer must reset whenever the status string changes (even between
// two pending states), not just when the boolean flips.
```

File: frontend/src/components/documents/PreviewStatePanel.jsx [hunk 9]
```diff
@@ -52,6 +67,17 @@ export function PreviewStatePanel({
           <p className="mb-4 text-sm text-gray-500">{formatBytes(documentData.file_size)}</p>
         ) : null}
         <p className="mb-6 text-gray-700">{message}</p>
+        {onRetry && (isFailed || (isPending && showStuckRetry)) && (
+          <div className="mb-6 -mt-2 text-sm text-gray-500">
+            Having trouble viewing?{' '}
+            <button
```
Comment (xiez) [inline frontend/src/components/documents/PreviewStatePanel.jsx:76]: 🟢 **[Low] Missing `type="button"` attribute**

Without an explicit `type`, a `<button>` element defaults to `type="submit"` in HTML. While this component isn't currently inside a `<form>`, the omission could cause accidental form submission if it ever is.

```jsx
<button
  type="button"
  onClick={onRetry}
  ...
>
```

File: backend/tests/documents/test_views.py [hunk 10]
```diff
@@ -959,6 +959,61 @@ def test_get_document_preview_data_too_many_pages(api_client, user):
     assert response.json()['render_error'] == "Document has too many pages to generate a preview."
 
 
+@pytest.mark.django_db
+@patch('documents.views.fileserver_client.delete_file')
+@patch('documents.services.generate_pdf_pages_task.delay')
+def test_get_document_preview_data_force_rebuild(mock_task_delay, mock_delete_file, api_client, user):
```
Comment (xiez) [inline backend/tests/documents/test_views.py]: 🟡 **[Medium] Missing boundary test: `force_rebuild` on a version with zero existing pages**

This test covers the happy path where pages already exist and need to be deleted. There is no test for the boundary case where `primary_version.pages` is already empty (e.g. a first-time render failure before any pages were written).

In that case `page_storage_keys` is `[]`, `mock_delete_file` should not be called, and the task should still be re-triggered. Per Rule 3 in `tech_stack_guidelines.md`, boundary conditions require dedicated tests.

Suggested test name: `test_get_document_preview_data_force_rebuild_no_existing_pages`

File: backend/documents/views.py (hunk 5 above)
Comment (xiez) [inline backend/documents/views.py:664]: 🔴 **[Critical] Stale in-flight Celery task can race against the newly-queued one**

If `generate_pdf_pages_task` was already running (e.g. `render_status == PROCESSING`) when `force_rebuild` fires, the force rebuild:
1. Deletes pages the running task is currently writing
2. Resets `has_pages=False`, so the already-running task now passes its `if version.has_pages: return` guard and continues
3. Enqueues a second concurrent task

Two tasks then race to write pages and update render status, producing corrupt or duplicated `DocumentPage` rows.

**Fix:** Check `render_status` before allowing a rebuild and reject if already processing:
```python
if force_rebuild and primary_version.render_status == DocumentVersion.RENDER_PROCESSING:
    return Response({'detail': 'Render already in progress, please wait.'}, status=status.HTTP_409_CONFLICT)
```

**Review by xiez — COMMENTED** [review]
## Code Review — PR #268: fix(preview) indirect object JSON serialization + preview retry trigger

The PDF indirect-object resolution logic is well-thought-out and the circular reference guard is a nice safety net. The frontend retry UX is clean. However there are two critical backend correctness issues that need addressing before merge.

### 🔴 Critical (must fix)
1. **`force_rebuild` is a GET-triggered destructive mutation** — violates HTTP idempotency, cannot be CSRF-protected, and can be triggered by browser prefetch or accidental reload.
2. **Stale Celery task race** — if a render is already in-flight when `force_rebuild` fires, two tasks race to write pages concurrently, producing corrupt `DocumentPage` rows.

### 🟡 Medium (should fix)
3. **No rate-limit/cooldown guard** — rapid calls to `?force_rebuild=true` spawn unbounded Celery tasks and repeatedly destroy mid-generation work.
4. **Missing `'updated_at'` in `update_fields`** — inconsistent with every other `version.save()` call in the codebase.
5. **`rebuildTriggerCount` reset race** — two `useEffect` hooks sharing the same deps create a fragile timing dependency on React's effect ordering.
6. **`setShowStuckRetry` timer dependency is correct but undocumented** — a future maintainer could simplify the dep array and break the per-status reset.
7. **`resolve_pdf_object` should be `_resolve_pdf_object`** — private helper exposed as public API by missing leading underscore.
8. **Missing boundary test** for `force_rebuild` on a version with zero existing pages.

### 🟢 Low (nit)
9. **`str(uri_obj)` fallback is too broad** — pypdf `NameObject`/`ArrayObject` would produce malformed URLs.
10. **Missing `type="button"` on retry `<button>`** — defaults to `submit`, unsafe inside a future form context.
11. **`id()` cycle detection deserves a comment** explaining the CPython address-reuse limitation.
